### PR TITLE
Feature/bonjour

### PR DESCRIPTION
**Pull request by doronz88**

### CODE REVIEW

File: pymobiledevice3/bonjour.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class BonjourListener(ServiceListener):
+    def __init__(self, pair_records: List[Mapping] = None):
+        super().__init__()
+
+        if pair_records is None:
+            pair_records = []
+
+        self.pair_records = pair_records
```
Comment (matan1008) [inline pymobiledevice3/bonjour.py]: merge with the if statement before

File: pymobiledevice3/bonjour.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        lockdown = None
+        try:
+            lockdown = LockdownClient(hostname=ipv4[0], autopair=False)
+
+            for pair_record in self.pair_records:
+                lockdown = LockdownClient(hostname=ipv4[0], autopair=False, pair_record=pair_record)
+                if lockdown.paired:
+                    break
+        except ConnectionRefusedError:
+            # handle device disconnects
```
Comment (matan1008) [inline pymobiledevice3/bonjour.py]: Don't add device on failures

File: pymobiledevice3/cli/bonjour.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import logging
+import plistlib
+from pathlib import Path
+
+import click
+
+from pymobiledevice3.bonjour import browse
+from pymobiledevice3.cli.cli_common import print_json
+
+logger = logging.getLogger(__name__)
```
Comment (matan1008) [inline pymobiledevice3/cli/bonjour.py]: Remove

File: tests/services/test_lockdown.py [hunk 4]
```diff
@@ -12,3 +15,8 @@ def test_lockdown_reconnect(lockdown):
     d2 = lockdown.date
 
     assert d1 < d2
+
+
+def test_bonjour(lockdown):
```
Comment (matan1008) [inline tests/services/test_lockdown.py]: Move to test_bonjour

File: pymobiledevice3/bonjour.py (hunk 1 above)
Comment (doronz88) [reply pymobiledevice3/bonjour.py]: done

File: pymobiledevice3/bonjour.py (hunk 2 above)
Comment (doronz88) [reply pymobiledevice3/bonjour.py]: done

File: tests/services/test_lockdown.py (hunk 4 above)
Comment (doronz88) [reply tests/services/test_lockdown.py]: done

File: pymobiledevice3/bonjour.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            'mac_address': self.mac_address,
+            'ipv4': self.ipv4,
+            'ipv6': self.ipv6,
+            'lockdown_info': self.lockdown.all_values
+        }
+
+
+class BonjourListener(ServiceListener):
+    def __init__(self, pair_records: List[Mapping] = None):
+        super().__init__()
```
Comment (yotamolenik) [inline pymobiledevice3/bonjour.py:36]: ServiceListener doesn't have a constructor.  
Does invoking super().__init__() do anything here?

File: pymobiledevice3/bonjour.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+        try:
+            lockdown = LockdownClient(hostname=ipv4[0], autopair=False)
+
+            for pair_record in self.pair_records:
+                lockdown = LockdownClient(hostname=ipv4[0], autopair=False, pair_record=pair_record)
+                if lockdown.paired:
+                    break
+        except ConnectionRefusedError:
+            return
```
Comment (yotamolenik) [inline pymobiledevice3/bonjour.py:62]: there could be a case where the debug prints are false. with a non-valid hostname `add_service` will not add anything but will still print "service x added".  
so maybe its fine and maybe we want one more debug print

File: tests/services/test_bonjour.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+from pymobiledevice3.bonjour import browse
+
+BROWSE_TIMEOUT = 1
```
Comment (yotamolenik) [inline tests/services/test_bonjour.py:3]: the default value for browse timeout is 5, but in the test you do it faster?

**Review by yotamolenik — APPROVED** [review]

**Pull request merged**